### ISSUE TITLE
Rename file from solution explorer might loose changes and undo doesn't work

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/project-system#4538]
Rename file from solution explorer might loose changes.

I noticed several times adding a new file, writing some code, then for whatever reason renaming the file, I end up with an empty class definition.

I just now do that with an existing file and choosing yes on the prompt to also rename the definition. It'll remove the change I just made. So this could also be the refactoring using an old version (from disk) to do the rename.

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/61984/rename-file-from-solution-explorer-might-loose-cha.html
VSTS ticketId: 441845_
_These are the original issue comments:_

Radames Cruz Moreno [MSFT] on 5/30/2017, 05:18 PM (615 days ago): <p>Thank you for your feedback! For us to investigate this further, could you please tell us:</p><ul><li>What type of project are you working on? Desktop application? Web application?</li><li>What programming language are you using?</li><li>How often does this happen?</li></ul><p><br></p><p>We look forward to hearing from you!</p>

Eric on 5/31/2017, 04:27 AM (614 days ago): <p>The projects I have been working on the most the last few days and which definitely have this problem are .NET Standard libraries in C#. Since they are meant to be used from UWP, they are targeting netstandard1.4. The solution also contains Xunit test projects and other console applications. I do have .NET Core 2.0 preview installed, but I'm not targeting it with this solution at least. But if I do, the same problems surface in the exact same way. See below.</p><p>Perhaps the title of the issue isn't quite correct. It is the rename refactoring that causes the issue. I can reproduce at will by hitting Ctrl+Shift+S, picking a file that has the same name as the class in it, writing a letter to make the file dirty, right clicking it, picking rename and appending say a number to the file name. The rename will take 10 seconds (another issue reported) but succeed, keeping the changes. I then get prompted to also rename the class and loose my changes if I respond with yes. The class is renamed, but the usual case is that it is again an empty class declaration, as this mostly happens after adding a new file. I do have Mads Kristensens Add New File extension installed and usually create files with Shift+F2, if that matters.</p><p>Trying just now to reproduce with dotnet new console, which automatically targets netcoreapp2.0 on my system, renaming Program to Program2 the issue surfaces. Renaming Program2 to Program, I do not get prompted again to rename the class Program2 back to Program. Trying again to rename to Program2, the file rename is very fast like it should, but apparently I won't be prompted again to refactor rename, which is an issue in itself as well. Until I save.</p><p>I find there are multiple problems here. It appears to go wrong in the way previously described only once for one file. Subsequent tries on the same file give different results, but it certainly doesn't work like it should. In both the larger solution and the dotnet new test in the same way, whichever file I pick to test with.</p><p>Sometimes the refactoring confirmation dialog does not appear until I save, I can even change the file between the rename and the save operation, but it will prompt me to rename when I then save. That's not quite how it should be either. I also find a warning dialog can appear, saying "The refactoring of code element 'ClassName' failed." with an old school yellow exclamation mark.</p><p>So just trying to randomly rename files, there is always an issue. I'm not renaming all the time, but I'd say it'd happen more than once every day that I loose 5-10 minutes work because of this.</p>

_These are the original issue solutions:_
(no solutions)
